### PR TITLE
Output Endpoint Implementation.

### PR DESCRIPTION
**Pull request by anu3990**
So we can incorporate your pull request, please share the following:
* What issue are you addressing with this pull request? https://project.marklogic.com/jira/browse/JAVA-90 and https://project.marklogic.com/jira/browse/JAVA-91
* Are you modifying the correct branch? develop-72
* Have you run unit tests? (See CONTRIBUTING.md)
* Version of MarkLogic Java Client API 5.0.1
* Version of MarkLogic Server (see admin gui on port 8001)
* Java version 11.0.4
* OS and version
* What Changed: What happened before this change? What happens without this change?

### CODE REVIEW

File: marklogic-client-api/src/main/java/com/marklogic/client/dataservices/impl/OutputEndpointImpl.java [hunk 1]
```diff
@@ -59,7 +62,59 @@ private BulkOutputCallerImpl(OutputEndpointImpl endpoint) {
...
+                    output = getEndpoint().getCaller().call(
+                            getEndpoint().getClient(), getEndpointState(), session, getWorkUnit()
+                    );
+                    incrementCallCount();
+                } catch(Throwable throwable) {
+                    throw new RuntimeException("error while calling "+getEndpoint().getEndpointPath(), throwable);
+                }
+
+                if (allowsEndpointState()) {
+                    if(output!= null && output.findFirst()!= null && output.findFirst().isPresent()) {
```
Comment (ehennum) [inline marklogic-client-api/src/main/java/com/marklogic/client/dataservices/impl/OutputEndpointImpl.java]: findFirst() is a terminal operation that completely consumes the stream.

Would it work to perform the tests in steps?

That is, if the output is not null, construct the array.  Then, if the array is not empty, process it?

File: marklogic-client-api/src/main/java/com/marklogic/client/dataservices/impl/OutputEndpointImpl.java [hunk 2]
```diff
@@ -59,7 +62,59 @@ private BulkOutputCallerImpl(OutputEndpointImpl endpoint) {
...
+
+                    output = getEndpoint().getCaller().call(
+                            getEndpoint().getClient(), getEndpointState(), session, getWorkUnit()
+                    );
+                    incrementCallCount();
+                } catch(Throwable throwable) {
+                    throw new RuntimeException("error while calling "+getEndpoint().getEndpointPath(), throwable);
+                }
+
+                if (allowsEndpointState()) {
```
Comment (ehennum) [inline marklogic-client-api/src/main/java/com/marklogic/client/dataservices/impl/OutputEndpointImpl.java]: Should the client also process the output if there is no endpoint state?

Would it work to move the construction of the result array before this if branch and the for loop after the if branch.

The if branch would also need to increment i from zero to 1.

File: marklogic-client-api/src/main/java/com/marklogic/client/dataservices/impl/OutputEndpointImpl.java [hunk 3]
```diff
@@ -59,7 +62,59 @@ private BulkOutputCallerImpl(OutputEndpointImpl endpoint) {
...
+                    if(output!= null && output.findFirst()!= null && output.findFirst().isPresent()) {
+                        InputStream[] result = output.toArray(size -> new InputStream[size]);
+                        setEndpointState(result[0]);
+                        for(int i=1; i<result.length; i++) {
+                            forEachOutput((Consumer<InputStream>) result[i]);
```
Comment (ehennum) [inline marklogic-client-api/src/main/java/com/marklogic/client/dataservices/impl/OutputEndpointImpl.java]: forEachOutput() is a setter for a Consumer function.  The loop should call the Consumer function's accept() method.

File: marklogic-client-api/src/main/java/com/marklogic/client/dataservices/impl/OutputEndpointImpl.java [hunk 4]
```diff
@@ -59,7 +62,59 @@ private BulkOutputCallerImpl(OutputEndpointImpl endpoint) {
         @Override
         public void awaitCompletion() {
```
Comment (ehennum) [inline marklogic-client-api/src/main/java/com/marklogic/client/dataservices/impl/OutputEndpointImpl.java]: Would it be appropriate to add a guard to make sure the output consumer is not null?

File: marklogic-client-api/src/main/java/com/marklogic/client/dataservices/impl/OutputEndpointImpl.java [hunk 5]
```diff
@@ -59,7 +62,59 @@ private BulkOutputCallerImpl(OutputEndpointImpl endpoint) {
...
             logger.trace("output endpoint running endpoint={} work={}", getEndpointPath(), getWorkUnit());
+            setPhase(WorkPhase.RUNNING);
```
Comment (ehennum) [inline marklogic-client-api/src/main/java/com/marklogic/client/dataservices/impl/OutputEndpointImpl.java]: Should we add a test that the current WorkPhase is INITIALIZING and throw an illegal state exception if it isn't?

File: marklogic-client-api/src/main/java/com/marklogic/client/dataservices/impl/OutputEndpointImpl.java [hunk 6]
```diff
@@ -61,7 +60,16 @@ private BulkOutputCallerImpl(OutputEndpointImpl endpoint) {
 
         @Override
         public void awaitCompletion() {
+            if(outputConsumer == null)
+                return;
```
Comment (ehennum) [inline marklogic-client-api/src/main/java/com/marklogic/client/dataservices/impl/OutputEndpointImpl.java]: Should it throw an exception instead?  If the application hasn't supplied a consumer, is it meaningful to await completion?

File: marklogic-client-api/src/main/java/com/marklogic/client/dataservices/impl/OutputEndpointImpl.java [hunk 7]
```diff
@@ -79,14 +87,17 @@ public void awaitCompletion() {
                     throw new RuntimeException("error while calling "+getEndpoint().getEndpointPath(), throwable);
                 }
 
+                InputStream[] result = null;
```
Comment (ehennum) [inline marklogic-client-api/src/main/java/com/marklogic/client/dataservices/impl/OutputEndpointImpl.java]: Should the result variable and all of the result processing move inside the branch that tests output is not null?  If there's no output, by definition, there's no result processing.

File: marklogic-client-api/src/main/java/com/marklogic/client/dataservices/impl/OutputEndpointImpl.java [hunk 8]
```diff
@@ -79,14 +87,17 @@ public void awaitCompletion() {
...
+                if(output != null) {
+                    result = output.toArray(size -> new InputStream[size]);
+                }
                 if (allowsEndpointState()) {
-                    if(output!= null && output.findFirst()!= null && output.findFirst().isPresent()) {
-                        InputStream[] result = output.toArray(size -> new InputStream[size]);
-                        setEndpointState(result[0]);
-                        for(int i=1; i<result.length; i++) {
-                            forEachOutput((Consumer<InputStream>) result[i]);
+                        if(result !=null) {
```
Comment (ehennum) [inline marklogic-client-api/src/main/java/com/marklogic/client/dataservices/impl/OutputEndpointImpl.java]: Should there be a test that the result.length is greater than 0?  Instead of nested if branches, would the code be easier to read as an && condition on the test of allowsEndpointState()?

File: marklogic-client-api/src/main/java/com/marklogic/client/dataservices/impl/OutputEndpointImpl.java [hunk 9]
```diff
@@ -79,14 +87,17 @@ public void awaitCompletion() {
...
+                            setEndpointState(result[0]);
                         }
-                    }
+                }
+                for(int i=1; i<result.length; i++) {
```
Comment (ehennum) [inline marklogic-client-api/src/main/java/com/marklogic/client/dataservices/impl/OutputEndpointImpl.java]: Should i be initialized to 0 if the allowsEndpointState() is false?  In that case, the first result would be output instead of the endpoint state.

File: marklogic-client-api/src/main/java/com/marklogic/client/dataservices/impl/OutputEndpointImpl.java [hunk 10]
```diff
@@ -87,17 +87,15 @@ public void awaitCompletion() {
...
 
-                InputStream[] result = null;
                 if(output != null) {
-                    result = output.toArray(size -> new InputStream[size]);
-                }
-                if (allowsEndpointState()) {
-                        if(result !=null) {
+                    InputStream[] result = output.toArray(size -> new InputStream[size]);
+
+                    if (allowsEndpointState() && result != null && result.length>0) {
```
Comment (ehennum) [inline marklogic-client-api/src/main/java/com/marklogic/client/dataservices/impl/OutputEndpointImpl.java]: As I understand Stream, if output is not null, the output.toArray() is guaranteed to be not null.

More generally, either the if test doesn't need to test for result != null or the for loop should have the same test.
